modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+        methodWriter.writeDebugInfo(irCastNode.getLocation());
+        methodWriter.writeCast(irCastNode.getCast());
+    }
+
+    @Override
+    public void visitInstanceof(InstanceofNode irInstanceofNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+
+        ExpressionNode irChildNode = irInstanceofNode.getChildNode();
+
+        visit(irChildNode, writeScope);
+
+        Class<?> instanceType = irInstanceofNode.getInstanceType();
+        Class<?> expressionType = irInstanceofNode.getExpressionType();
+
+        if (irInstanceofNode.getInstanceType() == def.class) {
+            methodWriter.writePop(MethodWriter.getType(expressionType).getSize());
+            methodWriter.push(true);
+        } else if (irChildNode.getExpressionType().isPrimitive()) {
+            methodWriter.writePop(MethodWriter.getType(expressionType).getSize());
+            methodWriter.push(PainlessLookupUtility.typeToBoxedType(instanceType).isAssignableFrom(
+                    PainlessLookupUtility.typeToBoxedType(irChildNode.getExpressionType())));
+        } else {
+            methodWriter.instanceOf(MethodWriter.getType(PainlessLookupUtility.typeToBoxedType(instanceType)));
+        }
+    }
+
+    @Override
+    public void visitConditional(ConditionalNode irConditionalNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irConditionalNode.getLocation());
+
+        Label fals = new Label();
+        Label end = new Label();
+
+        visit(irConditionalNode.getConditionNode(), writeScope);
+        methodWriter.ifZCmp(Opcodes.IFEQ, fals);
+
+        visit(irConditionalNode.getLeftNode(), writeScope);
+        methodWriter.goTo(end);
+        methodWriter.mark(fals);
+        visit(irConditionalNode.getRightNode(), writeScope);
+        methodWriter.mark(end);
+    }
+
+    @Override
+    public void visitElvis(ElvisNode irElvisNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irElvisNode.getLocation());
+
+        Label end = new Label();
+
+        visit(irElvisNode.getLeftNode(), writeScope);
+        methodWriter.dup();
+        methodWriter.ifNonNull(end);
+        methodWriter.pop();
+        visit(irElvisNode.getRightNode(), writeScope);
+        methodWriter.mark(end);
+    }
+
+    @Override
+    public void visitListInitialization(ListInitializationNode irListInitializationNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irListInitializationNode.getLocation());
+
+        PainlessConstructor painlessConstructor = irListInitializationNode.getConstructor();
+        methodWriter.newInstance(MethodWriter.getType(irListInitializationNode.getExpressionType()));
+        methodWriter.dup();
+        methodWriter.invokeConstructor(
```

Comment:
Pull these out into variables.

Reply:
Lots of overlap with naming so I'm going to leave this one.